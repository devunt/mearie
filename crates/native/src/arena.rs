use bumpalo::Bump;
use rustc_hash::FxHashSet;
use std::cell::RefCell;

pub struct Arena {
    bump: Bump,
    interned: RefCell<FxHashSet<&'static str>>,
}

impl Arena {
    #[inline]
    pub fn new() -> Self {
        Self {
            bump: Bump::new(),
            interned: RefCell::new(FxHashSet::default()),
        }
    }

    #[inline]
    pub fn allocator(&self) -> &Bump {
        &self.bump
    }

    #[inline]
    pub fn alloc<T>(&self, value: T) -> &T {
        self.bump.alloc(value)
    }

    #[inline]
    pub fn alloc_vec<T>(&self) -> bumpalo::collections::Vec<'_, T> {
        bumpalo::collections::Vec::new_in(&self.bump)
    }

    pub fn intern(&self, s: &str) -> &str {
        let mut interned = self.interned.borrow_mut();

        if let Some(&existing) = interned.get(s) {
            return existing;
        }

        let allocated = self.bump.alloc_str(s);

        // SAFETY: This lifetime extension from the arena's lifetime to 'static is sound because:
        // 1. The memory is allocated in the bump allocator owned by this Arena
        // 2. The bump allocator never moves or frees allocated memory until the Arena is dropped
        // 3. The string is only stored internally in the FxHashSet with a 'static lifetime
        // 4. The public API returns &str with the Arena's lifetime (not 'static), maintaining proper borrow checking
        // 5. The Arena's Drop implementation will clean up all memory, including this string
        //
        // This pattern is safe and commonly used in arena allocators for string interning.
        let static_str = unsafe { std::mem::transmute::<&str, &'static str>(allocated) };

        interned.insert(static_str);
        static_str
    }
}

impl Default for Arena {
    #[inline]
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_allocation() {
        let arena = Arena::new();

        let num = arena.alloc(42);
        assert_eq!(*num, 42);
    }

    #[test]
    fn test_string_interning_returns_same_pointer() {
        let arena = Arena::new();

        let s1 = arena.intern("hello");
        let s2 = arena.intern("hello");

        assert_eq!(s1, s2);
        assert!(std::ptr::eq(s1, s2));
    }

    #[test]
    fn test_different_strings_have_different_pointers() {
        let arena = Arena::new();

        let s1 = arena.intern("hello");
        let s2 = arena.intern("world");

        assert_ne!(s1, s2);
        assert!(!std::ptr::eq(s1, s2));
    }

    #[test]
    fn test_vec_allocation() {
        let arena = Arena::new();

        let mut vec = arena.alloc_vec();
        vec.push(1);
        vec.push(2);
        vec.push(3);

        assert_eq!(vec.len(), 3);
        assert_eq!(vec[0], 1);
        assert_eq!(vec[2], 3);
    }

    #[test]
    fn test_multiple_interned_strings() {
        let arena = Arena::new();

        let type1 = arena.intern("User");
        let type2 = arena.intern("Post");
        let type3 = arena.intern("User");

        assert!(std::ptr::eq(type1, type3));
        assert!(!std::ptr::eq(type1, type2));
    }

    #[test]
    fn test_complex_nested_structure() {
        let arena = Arena::new();

        let outer = arena.alloc(vec![
            arena.intern("field1"),
            arena.intern("field2"),
            arena.intern("field1"),
        ]);

        assert_eq!(outer.len(), 3);
        assert!(std::ptr::eq(outer[0], outer[2]));
    }
}
