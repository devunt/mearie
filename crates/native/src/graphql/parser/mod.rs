pub mod lexer;
#[allow(clippy::module_inception)]
pub mod parser;
pub mod state;

pub use parser::Parser;
pub use state::{Complete, Failed, Parsing, Ready, Uninitialized};
