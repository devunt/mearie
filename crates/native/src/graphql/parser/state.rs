/// Type-state marker indicating the parser has been created but not initialized with an arena.
pub struct Uninitialized;

/// Type-state marker indicating the parser has been initialized and is ready to parse.
pub struct Ready;

/// Type-state marker indicating the parser is currently parsing a document.
pub struct Parsing;

/// Type-state marker indicating the parser has successfully completed parsing.
pub struct Complete;

/// Type-state marker indicating the parser encountered an error during parsing.
pub struct Failed;
