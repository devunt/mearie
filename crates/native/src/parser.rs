pub mod context;
pub mod lexer;
pub mod parse;
pub mod printer;

pub use context::GraphQLContext;
pub use parse::ParseNode;
