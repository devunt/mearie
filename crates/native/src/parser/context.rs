use bumpalo::{Bump, collections::Vec};

pub struct GraphQLContext {
    alloc: Bump,
}

impl GraphQLContext {
    #[inline]
    pub fn new() -> Self {
        Self { alloc: Bump::new() }
    }

    #[inline]
    pub fn alloc<T>(&self, item: T) -> &T {
        self.alloc.alloc(item)
    }

    #[inline]
    pub fn alloc_vec<T>(&self) -> Vec<'_, T> {
        Vec::new_in(&self.alloc)
    }

    #[inline]
    pub(crate) fn allocator(&self) -> &Bump {
        &self.alloc
    }
}

impl Default for GraphQLContext {
    #[inline]
    fn default() -> Self {
        Self::new()
    }
}
