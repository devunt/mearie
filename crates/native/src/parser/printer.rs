use crate::ast::*;
use std::{fmt, fmt::Write};

pub trait PrintNode {
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result;

    fn print(&self) -> String {
        let mut buf = String::new();
        match self.write_to_buffer(0, &mut buf) {
            Ok(()) => buf,
            Err(_) => String::new(),
        }
    }
}

impl fmt::Display for dyn PrintNode {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.write_to_buffer(0, f)
    }
}

const INDENT_SIZE: usize = 2;

#[inline(always)]
fn write_indent(level: usize, buffer: &mut dyn Write) -> fmt::Result {
    for _ in 0..(level * INDENT_SIZE) {
        buffer.write_char(' ')?;
    }
    Ok(())
}

#[inline]
fn write_arguments<'a>(arguments: &[Argument<'a>], level: usize, buffer: &mut dyn Write) -> fmt::Result {
    let mut first = true;
    for arg in arguments {
        if !first {
            buffer.write_str(", ")?;
        }
        first = false;
        write!(buffer, "{}: ", arg.name)?;
        arg.value.write_to_buffer(level, buffer)?;
    }
    Ok(())
}

#[inline]
fn write_directives<'a>(directives: &[Directive<'a>], level: usize, buffer: &mut dyn Write) -> fmt::Result {
    for directive in directives {
        buffer.write_str(" @")?;
        buffer.write_str(&directive.name)?;
        if !directive.arguments.is_empty() {
            buffer.write_char('(')?;
            write_arguments(&directive.arguments, level, buffer)?;
            buffer.write_char(')')?;
        }
    }
    Ok(())
}

impl<'a> PrintNode for Value<'a> {
    #[inline]
    fn write_to_buffer(&self, _level: usize, buffer: &mut dyn Write) -> fmt::Result {
        match self {
            Value::Variable(name) => write!(buffer, "${}", name),
            Value::Int(value) => buffer.write_str(value),
            Value::Float(value) => buffer.write_str(value),
            Value::String(value) => {
                buffer.write_char('"')?;
                for c in value.chars() {
                    match c {
                        '\r' => buffer.write_str(r"\r")?,
                        '\n' => buffer.write_str(r"\n")?,
                        '\t' => buffer.write_str(r"\t")?,
                        '"' => buffer.write_str("\\\"")?,
                        '\\' => buffer.write_str(r"\\")?,
                        '\u{0020}'..='\u{FFFF}' => buffer.write_char(c)?,
                        _ => write!(buffer, "\\u{:04x}", c as u32)?,
                    }
                }
                buffer.write_char('"')
            }
            Value::Boolean(true) => buffer.write_str("true"),
            Value::Boolean(false) => buffer.write_str("false"),
            Value::Null => buffer.write_str("null"),
            Value::Enum(value) => buffer.write_str(value),
            Value::List(values) => {
                buffer.write_str("[")?;
                let mut first = true;
                for value in values.iter() {
                    if first {
                        first = false;
                    } else {
                        buffer.write_str(", ")?;
                    }
                    value.write_to_buffer(_level, buffer)?;
                }
                buffer.write_str("]")
            }
            Value::Object(fields) => {
                buffer.write_str("{")?;
                let mut first = true;
                for field in fields.iter() {
                    if first {
                        first = false;
                    } else {
                        buffer.write_str(", ")?;
                    }
                    write!(buffer, "{}: ", field.name)?;
                    field.value.write_to_buffer(_level, buffer)?;
                }
                buffer.write_str("}")
            }
        }
    }
}

impl<'a> PrintNode for Type<'a> {
    #[inline]
    fn write_to_buffer(&self, _level: usize, buffer: &mut dyn Write) -> fmt::Result {
        match self {
            Type::Named(named) => buffer.write_str(named.name),
            Type::List(inner) => {
                buffer.write_str("[")?;
                inner.write_to_buffer(_level, buffer)?;
                buffer.write_str("]")
            }
            Type::NonNull(inner) => match inner {
                NonNullType::Named(named) => {
                    buffer.write_str(named.name)?;
                    buffer.write_str("!")
                }
                NonNullType::List(inner) => {
                    buffer.write_str("[")?;
                    inner.write_to_buffer(_level, buffer)?;
                    buffer.write_str("]!")
                }
            },
        }
    }
}

impl<'a> PrintNode for SelectionSet<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        if !self.is_empty() {
            let level = level + 1;
            buffer.write_str("{")?;
            for selection in self.selections.iter() {
                buffer.write_char('\n')?;
                write_indent(level, buffer)?;
                match selection {
                    Selection::Field(field) => field.write_to_buffer(level, buffer)?,
                    Selection::FragmentSpread(spread) => spread.write_to_buffer(level, buffer)?,
                    Selection::InlineFragment(inline) => inline.write_to_buffer(level, buffer)?,
                };
            }
            buffer.write_char('\n')?;
            write_indent(level - 1, buffer)?;
            buffer.write_char('}')
        } else {
            Ok(())
        }
    }
}

impl<'a> PrintNode for Field<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        if let Some(alias) = self.alias {
            write!(buffer, "{}: {}", alias, self.name)?;
        } else {
            buffer.write_str(&self.name)?;
        }

        if !self.arguments.is_empty() {
            buffer.write_char('(')?;
            write_arguments(&self.arguments, level, buffer)?;
            buffer.write_char(')')?;
        }

        write_directives(&self.directives, level, buffer)?;

        if !self.selection_set.is_empty() {
            buffer.write_str(" ")?;
        }
        self.selection_set.write_to_buffer(level, buffer)
    }
}

impl<'a> PrintNode for FragmentSpread<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        write!(buffer, "...{}", self.fragment_name)?;
        write_directives(&self.directives, level, buffer)
    }
}

impl<'a> PrintNode for InlineFragment<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        buffer.write_str("...")?;
        if let Some(type_condition) = self.type_condition {
            write!(buffer, " on {}", type_condition)?;
        }
        write_directives(&self.directives, level, buffer)?;
        buffer.write_char(' ')?;
        self.selection_set.write_to_buffer(level, buffer)
    }
}

impl<'a> PrintNode for VariableDefinition<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        write!(buffer, "${}: ", self.variable)?;
        self.typ.write_to_buffer(level, buffer)?;
        if let Some(default_value) = &self.default_value {
            buffer.write_str(" = ")?;
            default_value.write_to_buffer(level, buffer)?;
        }
        write_directives(&self.directives, level, buffer)
    }
}

impl<'a> PrintNode for OperationDefinition<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        if self.operation_type == OperationType::Query
            && self.name.is_none()
            && self.variable_definitions.is_empty()
            && self.directives.is_empty()
        {
            return self.selection_set.write_to_buffer(level, buffer);
        }

        match self.operation_type {
            OperationType::Query => buffer.write_str("query")?,
            OperationType::Mutation => buffer.write_str("mutation")?,
            OperationType::Subscription => buffer.write_str("subscription")?,
        }

        if let Some(name) = self.name {
            write!(buffer, " {}", name)?;
        }

        if !self.variable_definitions.is_empty() {
            if self.name.is_none() {
                buffer.write_char(' ')?;
            }
            buffer.write_char('(')?;
            let mut first = true;
            for var_def in &self.variable_definitions {
                if !first {
                    buffer.write_str(", ")?;
                }
                first = false;
                var_def.write_to_buffer(level, buffer)?;
            }
            buffer.write_char(')')?;
        }

        write_directives(&self.directives, level, buffer)?;

        buffer.write_str(" ")?;
        self.selection_set.write_to_buffer(level, buffer)
    }
}

impl<'a> PrintNode for FragmentDefinition<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        write!(buffer, "fragment {} on {}", self.name, self.type_condition)?;
        write_directives(&self.directives, level, buffer)?;
        buffer.write_char(' ')?;
        self.selection_set.write_to_buffer(level, buffer)
    }
}

impl<'a> PrintNode for ExecutableDefinition<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        match self {
            ExecutableDefinition::Operation(op) => op.write_to_buffer(level, buffer),
            ExecutableDefinition::Fragment(frag) => frag.write_to_buffer(level, buffer),
        }
    }
}

impl<'a> PrintNode for Document<'a> {
    #[inline]
    fn write_to_buffer(&self, level: usize, buffer: &mut dyn Write) -> fmt::Result {
        let mut first = true;
        for definition in self.definitions.iter() {
            if first {
                first = false;
            } else {
                buffer.write_str("\n\n")?;
            }
            if let Definition::Executable(exec) = definition {
                exec.write_to_buffer(level, buffer)?
            }
        }
        Ok(())
    }

    #[inline]
    fn print(&self) -> String {
        let capacity = self.source.code.len();
        let mut buf = String::with_capacity(capacity);
        match self.write_to_buffer(0, &mut buf) {
            Ok(()) => buf,
            Err(_) => String::new(),
        }
    }
}
