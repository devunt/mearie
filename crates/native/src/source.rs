use serde::{Deserialize, Serialize};

#[cfg_attr(feature = "napi", napi_derive::napi(object))]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SourceBuf {
    pub code: String,
    pub file_path: String,
    pub start_line: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Source<'a> {
    pub code: &'a str,
    pub file_path: &'a str,
    pub start_line: u32,
}

impl<'a> Source<'a> {
    pub fn ephemeral(code: &'a str) -> Self {
        Source {
            code,
            file_path: "test.graphql",
            start_line: 1,
        }
    }
}

impl<'a> From<&'a SourceBuf> for Source<'a> {
    fn from(owned: &'a SourceBuf) -> Self {
        Source {
            code: &owned.code,
            file_path: &owned.file_path,
            start_line: owned.start_line,
        }
    }
}
