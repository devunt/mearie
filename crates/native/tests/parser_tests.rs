use insta::assert_debug_snapshot;
use mearie_native::arena::Arena;
use mearie_native::graphql::parser::Parser;
use mearie_native::source::Source;

macro_rules! parse {
    ($code:expr) => {{
        let arena = Box::leak(Box::new(Arena::new()));
        let source = Box::leak(Box::new(Source::ephemeral($code)));
        Parser::new(arena).with_source(source).parse().unwrap()
    }};
}

// =============================================================================
// QUERIES
// =============================================================================

#[test]
fn test_simple_query() {
    assert_debug_snapshot!(parse!("query { user { id name } }"));
}

#[test]
fn test_query_with_variables() {
    assert_debug_snapshot!(parse!("query GetUser($id: ID!) { user(id: $id) { id name email } }"));
}

#[test]
fn test_query_with_default_values() {
    assert_debug_snapshot!(parse!(
        "query GetUsers($limit: Int = 10, $offset: Int = 0) { users(limit: $limit, offset: $offset) { id } }"
    ));
}

#[test]
fn test_query_with_variables_and_defaults() {
    assert_debug_snapshot!(parse!(
        r#"query GetUsers(
            $limit: Int = 10
            $offset: Int = 0
            $sortBy: String = "name"
            $ascending: Boolean = true
        ) {
            users(limit: $limit, offset: $offset, sortBy: $sortBy, ascending: $ascending) {
                id
                name
            }
        }"#
    ));
}

#[test]
fn test_query_with_directives() {
    assert_debug_snapshot!(parse!(
        "query GetUser($withEmail: Boolean!) { user { id name email @include(if: $withEmail) } }"
    ));
}

#[test]
fn test_query_with_skip_and_include() {
    assert_debug_snapshot!(parse!(
        r#"query GetUser($withEmail: Boolean!, $withoutPosts: Boolean!) {
            user {
                id
                name
                email @include(if: $withEmail)
                posts @skip(if: $withoutPosts) { id }
            }
        }"#
    ));
}

#[test]
fn test_query_with_field_directives() {
    assert_debug_snapshot!(parse!("query { user @skip(if: true) { id } }"));
}

#[test]
fn test_query_with_aliases() {
    assert_debug_snapshot!(parse!("query { userId: user { userName: name } }"));
}

#[test]
fn test_nested_selection_sets() {
    assert_debug_snapshot!(parse!("query { user { posts { comments { author { name } } } } }"));
}

#[test]
fn test_complex_nested_query() {
    assert_debug_snapshot!(parse!(
        "query { user { posts { comments { author { posts { id } } } } } }"
    ));
}

#[test]
fn test_query_with_list_arguments() {
    assert_debug_snapshot!(parse!(
        r#"query { users(ids: [1, 2, 3], tags: ["admin", "user"]) { id } }"#
    ));
}

#[test]
fn test_query_with_object_arguments() {
    assert_debug_snapshot!(parse!(
        r#"query { createUser(input: { name: "John", age: 30, tags: ["admin"] }) { id } }"#
    ));
}

#[test]
fn test_query_with_complex_nesting() {
    assert_debug_snapshot!(parse!(
        r#"query ComplexQuery($userId: ID!, $includeComments: Boolean = false) {
            user(id: $userId) {
                id
                name
                posts(first: 10) {
                    edges {
                        node {
                            id
                            title
                            comments(first: 5) @include(if: $includeComments) {
                                edges {
                                    node {
                                        id
                                        content
                                        author {
                                            id
                                            name
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }"#
    ));
}

#[test]
fn test_multiple_operations() {
    assert_debug_snapshot!(parse!(
        r#"query GetUser { user { id } }
        query GetPosts { posts { id } }
        mutation CreatePost { createPost { id } }"#
    ));
}

// =============================================================================
// MUTATIONS
// =============================================================================

#[test]
fn test_simple_mutation() {
    assert_debug_snapshot!(parse!(r#"mutation CreateUser { createUser(name: "John") { id } }"#));
}

#[test]
fn test_mutation_with_input() {
    assert_debug_snapshot!(parse!(
        "mutation CreateUser($input: UserInput!) { createUser(input: $input) { id } }"
    ));
}

#[test]
fn test_mutation_with_complex_input() {
    assert_debug_snapshot!(parse!(
        r#"mutation CreatePost($input: PostInput!) {
            createPost(input: $input) {
                id
                title
                author { id name }
            }
        }"#
    ));
}

// =============================================================================
// SUBSCRIPTIONS
// =============================================================================

#[test]
fn test_subscription() {
    assert_debug_snapshot!(parse!("subscription { messageAdded { id content author { name } } }"));
}

// =============================================================================
// FRAGMENTS
// =============================================================================

#[test]
fn test_named_fragment() {
    assert_debug_snapshot!(parse!("fragment UserFields on User { id name email }"));
}

#[test]
fn test_query_with_fragment() {
    assert_debug_snapshot!(parse!(
        "query { user { ...UserFields } } fragment UserFields on User { id name email }"
    ));
}

#[test]
fn test_query_with_fragment_spread() {
    assert_debug_snapshot!(parse!(
        r#"query { user { ...UserFields } }
        fragment UserFields on User { id name }"#
    ));
}

#[test]
fn test_inline_fragment() {
    assert_debug_snapshot!(parse!("query { search { ... on User { name } } }"));
}

#[test]
fn test_nested_fragments() {
    assert_debug_snapshot!(parse!(
        r#"query { user { ...UserWithPosts } }
        fragment UserWithPosts on User { ...UserBasic posts { id } }
        fragment UserBasic on User { id name }"#
    ));
}

#[test]
fn test_nested_fragment_spreads() {
    assert_debug_snapshot!(parse!(
        r#"query { user { ...UserWithPosts } }
        fragment UserWithPosts on User { id ...UserBasic }
        fragment UserBasic on User { name email }"#
    ));
}

#[test]
fn test_fragment_with_nested_spreads() {
    assert_debug_snapshot!(parse!(
        r#"query { user { ...UserComplete } }
        fragment UserComplete on User { ...UserBasic ...UserProfile }
        fragment UserBasic on User { id name }
        fragment UserProfile on User { email avatar }"#
    ));
}

#[test]
fn test_query_with_inline_and_named_fragments() {
    assert_debug_snapshot!(parse!(
        r#"query {
            search {
                ...ResultFields
                ... on User { email }
                ... on Post { content }
            }
        }
        fragment ResultFields on SearchResult { id }"#
    ));
}

#[test]
fn test_query_with_fragments_and_variables() {
    assert_debug_snapshot!(parse!(
        r#"query GetUser($id: ID!, $withPosts: Boolean!) {
            user(id: $id) {
                ...UserFields
                posts @include(if: $withPosts) { ...PostFields }
            }
        }
        fragment UserFields on User { id name email }
        fragment PostFields on Post { id title content }"#
    ));
}

// =============================================================================
// TYPE DEFINITIONS
// =============================================================================

// Object Types

#[test]
fn test_object_type() {
    assert_debug_snapshot!(parse!("type User { id: ID! name: String! email: String }"));
}

#[test]
fn test_complete_schema() {
    assert_debug_snapshot!(parse!(
        "type Query { user(id: ID!): User users: [User!]! } type User { id: ID! name: String! email: String! }"
    ));
}

#[test]
fn test_type_with_implements() {
    assert_debug_snapshot!(parse!("type User implements Node { id: ID! }"));
}

#[test]
fn test_type_with_multiple_interfaces() {
    assert_debug_snapshot!(parse!("type User implements Node & Timestamped { id: ID! }"));
}

#[test]
fn test_type_implementing_multiple_interfaces() {
    assert_debug_snapshot!(parse!(
        r#"interface Node { id: ID! }
        interface Timestamped { createdAt: String! updatedAt: String! }
        type User implements Node & Timestamped {
            id: ID!
            createdAt: String!
            updatedAt: String!
            name: String!
        }"#
    ));
}

#[test]
fn test_type_with_description() {
    assert_debug_snapshot!(parse!(r#""A user in the system" type User { id: ID! }"#));
}

#[test]
fn test_type_with_field_descriptions() {
    assert_debug_snapshot!(parse!(
        r#""A user in the system"
        type User {
            "The user's unique identifier"
            id: ID!
            "The user's full name"
            name: String!
        }"#
    ));
}

#[test]
fn test_field_with_description() {
    assert_debug_snapshot!(parse!(r#"type User { "User ID" id: ID! }"#));
}

#[test]
fn test_field_with_arguments() {
    assert_debug_snapshot!(parse!("type Query { users(first: Int, offset: Int): [User] }"));
}

#[test]
fn test_field_with_multiple_arguments() {
    assert_debug_snapshot!(parse!(
        r#"type Query {
            users(
                first: Int
                after: String
                orderBy: OrderBy
                filter: UserFilter
            ): [User!]!
        }
        type User { id: ID! }"#
    ));
}

// Interfaces

#[test]
fn test_interface_definition() {
    assert_debug_snapshot!(parse!("interface Node { id: ID! createdAt: String }"));
}

#[test]
fn test_interface_with_fields() {
    assert_debug_snapshot!(parse!("interface Node { id: ID! createdAt: String! }"));
}

#[test]
fn test_schema_with_interface() {
    assert_debug_snapshot!(parse!(
        "interface Node { id: ID! } type User implements Node { id: ID! name: String! }"
    ));
}

#[test]
fn test_interface_with_multiple_implementations() {
    assert_debug_snapshot!(parse!(
        r#"interface Node { id: ID! }
        type User implements Node { id: ID! name: String! }
        type Post implements Node { id: ID! title: String! }
        type Comment implements Node { id: ID! content: String! }"#
    ));
}

// Unions

#[test]
fn test_union_definition() {
    assert_debug_snapshot!(parse!("union SearchResult = User | Post | Comment"));
}

#[test]
fn test_schema_with_union() {
    assert_debug_snapshot!(parse!(
        "union SearchResult = User | Post type User { id: ID! } type Post { id: ID! }"
    ));
}

#[test]
fn test_union_with_four_types() {
    assert_debug_snapshot!(parse!("union SearchResult = User | Post | Comment | Tag"));
}

// Enums

#[test]
fn test_enum_definition() {
    assert_debug_snapshot!(parse!("enum Status { ACTIVE INACTIVE PENDING }"));
}

#[test]
fn test_enum_with_description() {
    assert_debug_snapshot!(parse!(r#""User status" enum Status { ACTIVE INACTIVE }"#));
}

#[test]
fn test_enum_with_value_descriptions() {
    assert_debug_snapshot!(parse!(
        r#""User account status"
        enum Status {
            "Account is active"
            ACTIVE
            "Account is inactive"
            INACTIVE
        }"#
    ));
}

#[test]
fn test_schema_with_enum() {
    assert_debug_snapshot!(parse!(
        "enum Status { ACTIVE INACTIVE PENDING } type User { status: Status! }"
    ));
}

// Input Types

#[test]
fn test_input_definition() {
    assert_debug_snapshot!(parse!("input UserInput { name: String! email: String! age: Int }"));
}

#[test]
fn test_input_with_default_values() {
    assert_debug_snapshot!(parse!("input UserInput { name: String! age: Int = 18 }"));
}

#[test]
fn test_input_with_nested_types() {
    assert_debug_snapshot!(parse!(
        r#"input AddressInput { street: String! city: String! }
        input UserInput { name: String! address: AddressInput! }"#
    ));
}

#[test]
fn test_schema_with_input_type() {
    assert_debug_snapshot!(parse!(
        r#"input UserInput { name: String! email: String! age: Int }
        type Mutation { createUser(input: UserInput!): User }
        type User { id: ID! }"#
    ));
}

// Scalars

#[test]
fn test_scalar_definition() {
    assert_debug_snapshot!(parse!("scalar DateTime"));
}

#[test]
fn test_schema_with_custom_scalars() {
    assert_debug_snapshot!(parse!(
        "scalar DateTime scalar JSON type User { createdAt: DateTime! metadata: JSON }"
    ));
}

#[test]
fn test_all_scalar_types() {
    assert_debug_snapshot!(parse!(
        r#"type Example {
            id: ID!
            name: String!
            age: Int!
            price: Float!
            active: Boolean!
        }"#
    ));
}

// =============================================================================
// SCHEMA DEFINITIONS
// =============================================================================

#[test]
fn test_schema_definition() {
    assert_debug_snapshot!(parse!("schema { query: Query mutation: Mutation }"));
}

#[test]
fn test_schema_with_all_operations() {
    assert_debug_snapshot!(parse!(
        "schema { query: Query mutation: Mutation subscription: Subscription }"
    ));
}

#[test]
fn test_complete_schema_definition() {
    assert_debug_snapshot!(parse!(
        r#"schema {
            query: Query
            mutation: Mutation
            subscription: Subscription
        }
        type Query { user: User }
        type Mutation { createUser: User }
        type Subscription { userAdded: User }
        type User { id: ID! }"#
    ));
}

#[test]
fn test_large_schema() {
    assert_debug_snapshot!(parse!(
        r#"type Query {
            user(id: ID!): User
            users(first: Int, after: String): UserConnection!
            post(id: ID!): Post
            posts(first: Int, after: String): PostConnection!
        }
        type Mutation {
            createUser(input: UserInput!): User!
            updateUser(id: ID!, input: UserInput!): User!
            deleteUser(id: ID!): Boolean!
        }
        type Subscription {
            userAdded: User!
            userUpdated: User!
            userDeleted: ID!
        }
        type User implements Node {
            id: ID!
            name: String!
            email: String!
            posts(first: Int): PostConnection!
        }
        type Post implements Node {
            id: ID!
            title: String!
            content: String!
            author: User!
        }
        interface Node { id: ID! }
        type UserConnection {
            edges: [UserEdge!]!
            pageInfo: PageInfo!
        }
        type UserEdge {
            node: User!
            cursor: String!
        }
        type PostConnection {
            edges: [PostEdge!]!
            pageInfo: PageInfo!
        }
        type PostEdge {
            node: Post!
            cursor: String!
        }
        type PageInfo {
            hasNextPage: Boolean!
            hasPreviousPage: Boolean!
            startCursor: String
            endCursor: String
        }
        input UserInput { name: String! email: String! }"#
    ));
}

// =============================================================================
// DIRECTIVES
// =============================================================================

#[test]
fn test_directive_definition() {
    assert_debug_snapshot!(parse!("directive @auth(requires: Role!) on FIELD_DEFINITION"));
}

#[test]
fn test_schema_with_directive_definition() {
    assert_debug_snapshot!(parse!(
        "directive @auth(requires: Role!) on FIELD_DEFINITION | OBJECT enum Role { ADMIN USER }"
    ));
}

#[test]
fn test_repeatable_directive() {
    assert_debug_snapshot!(parse!(
        "directive @tag(name: String!) repeatable on FIELD_DEFINITION type User { id: ID! @tag(name: \"public\") @tag(name: \"searchable\") }"
    ));
}

#[test]
fn test_repeatable_directive_definition() {
    assert_debug_snapshot!(parse!("directive @tag(name: String!) repeatable on FIELD"));
}

// =============================================================================
// TYPE MODIFIERS
// =============================================================================

#[test]
fn test_list_type() {
    assert_debug_snapshot!(parse!("type User { tags: [String] }"));
}

#[test]
fn test_non_null_type() {
    assert_debug_snapshot!(parse!("type User { id: ID! name: String! }"));
}

#[test]
fn test_non_null_list() {
    assert_debug_snapshot!(parse!("type User { tags: [String!]! }"));
}

#[test]
fn test_list_and_non_null_combinations() {
    assert_debug_snapshot!(parse!(
        r#"type Example {
            nullableList: [String]
            nonNullList: [String]!
            listOfNonNulls: [String!]
            nonNullListOfNonNulls: [String!]!
        }"#
    ));
}

// =============================================================================
// SCHEMA EXTENSIONS
// =============================================================================

#[test]
fn test_extend_type() {
    assert_debug_snapshot!(parse!("extend type User { avatar: String }"));
}

#[test]
fn test_extend_type_with_base() {
    assert_debug_snapshot!(parse!("type User { id: ID! } extend type User { email: String! }"));
}

#[test]
fn test_extend_interface() {
    assert_debug_snapshot!(parse!("extend interface Node { createdAt: String }"));
}

#[test]
fn test_extend_interface_with_base() {
    assert_debug_snapshot!(parse!(
        "interface Node { id: ID! } extend interface Node { createdAt: String! }"
    ));
}

#[test]
fn test_extend_union() {
    assert_debug_snapshot!(parse!(
        "union SearchResult = User | Post extend union SearchResult = Comment"
    ));
}

#[test]
fn test_extend_enum() {
    assert_debug_snapshot!(parse!("enum Status { ACTIVE INACTIVE } extend enum Status { PENDING }"));
}

#[test]
fn test_extend_input() {
    assert_debug_snapshot!(parse!(
        "input UserInput { name: String! } extend input UserInput { email: String! }"
    ));
}
