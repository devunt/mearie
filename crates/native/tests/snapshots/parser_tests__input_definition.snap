---
source: crates/native/tests/parser_tests.rs
expression: "parse!(\"input UserInput { name: String! email: String! age: Int }\")"
---
Document {
    source: Source {
        code: "input UserInput { name: String! email: String! age: Int }",
        file_path: "test.graphql",
        start_line: 1,
    },
    definitions: [
        TypeSystem(
            Type(
                InputObject(
                    InputObjectTypeDefinition {
                        description: None,
                        name: TypeName(
                            Name(
                                "UserInput",
                            ),
                        ),
                        directives: [],
                        fields: [
                            InputValueDefinition {
                                description: None,
                                name: ArgumentName(
                                    Name(
                                        "name",
                                    ),
                                ),
                                typ: NonNull(
                                    Named(
                                        NamedType {
                                            name: TypeName(
                                                Name(
                                                    "String",
                                                ),
                                            ),
                                        },
                                    ),
                                ),
                                default_value: None,
                                directives: [],
                            },
                            InputValueDefinition {
                                description: None,
                                name: ArgumentName(
                                    Name(
                                        "email",
                                    ),
                                ),
                                typ: NonNull(
                                    Named(
                                        NamedType {
                                            name: TypeName(
                                                Name(
                                                    "String",
                                                ),
                                            ),
                                        },
                                    ),
                                ),
                                default_value: None,
                                directives: [],
                            },
                            InputValueDefinition {
                                description: None,
                                name: ArgumentName(
                                    Name(
                                        "age",
                                    ),
                                ),
                                typ: Named(
                                    NamedType {
                                        name: TypeName(
                                            Name(
                                                "Int",
                                            ),
                                        ),
                                    },
                                ),
                                default_value: None,
                                directives: [],
                            },
                        ],
                    },
                ),
            ),
        ),
    ],
}
